# EDA: Diagnosing Diabetes

In this project, you'll imagine you are a data scientist interested in exploring data that looks at how certain diagnostic factors affect the diabetes outcome of women patients.

You will use your EDA skills to help inspect, clean, and validate the data.

**Note**: This [dataset](https://www.kaggle.com/uciml/pima-indians-diabetes-database) is from the National Institute of Diabetes and Digestive and Kidney Diseases. It contains the following columns:

- `Pregnancies`: Number of times pregnant
- `Glucose`: Plasma glucose concentration per 2 hours in an oral glucose tolerance test
- `BloodPressure`: Diastolic blood pressure
- `SkinThickness`: Triceps skinfold thickness
- `Insulin`: 2-Hour serum insulin
- `BMI`: Body mass index
- `DiabetesPedigreeFunction`: Diabetes pedigree function
- `Age`: Age (years)
- `Outcome`: Class variable (0 or 1)

Let's get started!

## Initial Inspection

1. First, familiarize yourself with the dataset [here](https://www.kaggle.com/uciml/pima-indians-diabetes-database).

   Look at each of the nine columns in the documentation.
   
   What do you expect each data type to be?

Expected data type for each column:

- `Pregnancies`: `int64`
- `Glucose`: `int64`
- `BloodPressure`: `int64`
- `SkinThickness`: `int64`
- `Insulin`: `int64`
- `BMI`: `float64`
- `DiabetesPedigreeFunction`: `float64`
- `Age`: `int64`
- `Outcome`: `int64`

2. Next, let's load in the diabetes data to start exploring.

   Load the data in a variable called `diabetes_data` and print the first few rows.
   
   **Note**: The data is stored in a file called `diabetes.csv`.

In [1]:
import pandas as pd
import numpy as np

# load in data
diabetes_data = pd.read_csv('E:\\Semo4ka\\Python\\CodeCademy\\eda-diabetes\\diabetes.csv')
print(diabetes_data.head())

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age Outcome  
0                     0.627   50       1  
1                     0.351   31       0  
2                     0.672   32       1  
3                     0.167   21       0  
4                     2.288   33       1  


3. How many columns (features) does the data contain?

In [2]:
# print number of columns
print(len(diabetes_data.columns))

9


4. How many rows (observations) does the data contain?

In [3]:
# print number of rows
print(len(diabetes_data))

768


## Further Inspection

5. Let's inspect `diabetes_data` further.

   Do any of the columns in the data contain null (missing) values?

In [4]:
# find whether columns contain null values
print(diabetes_data.isnull().sum())

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


6. If you answered no to the question above, not so fast!

   While it's technically true that none of the columns contain null values, that doesn't necessarily mean that the data isn't missing any values.
   
   When exploring data, you should always question your assumptions and try to dig deeper.
   
   To investigate further, calculate summary statistics on `diabetes_data` using the `.describe()` method.

In [5]:
# perform summary statistics
print(diabetes_data.describe())

       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  768.000000                768.000000  768.000000  
mean    31.992578                  0.471876   33.240885  
std      7.884160                  0.331329   11.760232  
min      0.000000                  0.078000   21.000000  
25%     27.300000        

7. Looking at the summary statistics, do you notice anything odd about the following columns?

   - `Glucose`
   - `BloodPressure`
   - `SkinThickness`
   - `Insulin`
   - `BMI`

**Our response to question 7**: If you take a look at the minimum values for these five columns, you'll notice that they are all `0`. 

Blood Pressure or BMI can not be `0`! These values also seem to be way off from their respective medians and means, another indicator that something is off.

One way to interpret this is that there are missing values in the data.

8. Do you spot any other outliers in the data?

**Our response to question 8**: 

In addition to the `0` values that show up for the columns above, there appear to be additional outliers, such as:

- The maximum value of the `Insulin` column is `846`, which is abnormally high.
- The maximum value of the `Pregnancies` column is `17`. While having 17 pregnancies is not impossible, this case might be something to look further into to determine its accuracy.



9. Let's see if we can get a more accurate view of the missing values in the data.

   We will replace the instances of `0` with `NaN` in the five columns mentioned.

In [6]:
# replace instances of 0 with NaN
diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] =\
    diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, np.NaN)

10. Next, check for missing (null) values in all of the columns just like you did in Step 5.

    Now how many missing values are there?

In [7]:
# find whether columns contain null values after replacements are made
print(diabetes_data.isnull().sum())

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64


11. Let's take a closer look at these rows to get a better idea of _why_ some data might be missing.

    Print out all the rows that contain missing (null) values.

In [8]:
# print rows with missing values
print(diabetes_data[diabetes_data.isnull().any(axis=1)])

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0           35.0      NaN  33.6   
1              1     85.0           66.0           29.0      NaN  26.6   
2              8    183.0           64.0            NaN      NaN  23.3   
5              5    116.0           74.0            NaN      NaN  25.6   
7             10    115.0            NaN            NaN      NaN  35.3   
..           ...      ...            ...            ...      ...   ...   
761            9    170.0           74.0           31.0      NaN  44.0   
762            9     89.0           62.0            NaN      NaN  22.5   
764            2    122.0           70.0           27.0      NaN  36.8   
766            1    126.0           60.0            NaN      NaN  30.1   
767            1     93.0           70.0           31.0      NaN  30.4   

     DiabetesPedigreeFunction  Age Outcome  
0                       0.627   50       1  
1                    

12. Go through the rows with missing data. Do you notice any patterns or overlaps between the missing data?

**Our response to question 12**: 

We noticed that most rows with missing data have missing values in more than one column. In fact, every single row with at least one missing value also has a missing value in the `Insulin` column. This is a clue as to why the data is missing! If patients did not have their insulin measured, why might they also not have had these other measurements taken?

13. Next, take a closer look at the data types of each column in `diabetes_data`.

    Does the result match what you would expect?

In [9]:
# print data types using .info() method
print(diabetes_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   763 non-null    float64
 2   BloodPressure             733 non-null    float64
 3   SkinThickness             541 non-null    float64
 4   Insulin                   394 non-null    float64
 5   BMI                       757 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    object 
dtypes: float64(6), int64(2), object(1)
memory usage: 54.1+ KB
None


14. To figure out why the `Outcome` column is of type `object` (string) instead of type `int64`, print out the unique values in the `Outcome` column.

In [10]:
# print unique values of Outcome column
print(diabetes_data.Outcome.unique())

['1' '0' 'O']


15. How might you resolve this issue?

**Our response to question 15**: 

A possible next step would be to replace instances of `'O'` with `0` and convert the `Outcome` column to type `int64`.

In [11]:
diabetes_data.Outcome = diabetes_data.Outcome.replace({'O': 0})
diabetes_data.Outcome = pd.to_numeric(diabetes_data.Outcome)

## Next Steps:

16. Congratulations! In this project, you saw how EDA can help with the initial data inspection and cleaning process. This is an important step as it helps to keep your datasets clean and reliable.

    Here are some ways you might extend this project if you'd like:
    - Use `.value_counts()` to more fully explore the values in each column.
    - Investigate other outliers in the data that may be easily overlooked.
    - Instead of changing the `0` values in the five columns to `NaN`, try replacing the values with the median or mean of each column.

In [13]:
def val_and_counts(columns):
    for column in columns:
        print('Values and counts for {} column'.format(column))
        print(diabetes_data[column].value_counts().sort_index())
        
data_columns = diabetes_data.columns
val_and_counts(data_columns)

Values and counts for Pregnancies column
0     111
1     135
2     103
3      75
4      68
5      57
6      50
7      45
8      38
9      28
10     24
11     11
12      9
13     10
14      2
15      1
17      1
Name: Pregnancies, dtype: int64
Values and counts for Glucose column
44.0     1
56.0     1
57.0     2
61.0     1
62.0     1
        ..
195.0    2
196.0    3
197.0    4
198.0    1
199.0    1
Name: Glucose, Length: 135, dtype: int64
Values and counts for BloodPressure column
24.0      1
30.0      2
38.0      1
40.0      1
44.0      4
46.0      2
48.0      5
50.0     13
52.0     11
54.0     11
55.0      2
56.0     12
58.0     21
60.0     37
61.0      1
62.0     34
64.0     43
65.0      7
66.0     30
68.0     45
70.0     57
72.0     44
74.0     52
75.0      8
76.0     39
78.0     45
80.0     40
82.0     30
84.0     23
85.0      6
86.0     21
88.0     25
90.0     22
92.0      8
94.0      6
95.0      1
96.0      4
98.0      3
100.0     3
102.0     1
104.0     2
106.0     3
108.0     2

By exploring `SkinThickness` unique values, we found value of 99, wich is not realistic, so we can remove it from our data.

Also we will be removing `Pregnancies` with values 14 and abowe.

Next we will replace missing `Insulin` values with data's average.

In [17]:
print(diabetes_data.SkinThickness.unique())
diabetes_data = diabetes_data[diabetes_data.SkinThickness != 99]

preg_val_to_delete = [14, 15, 17]
diabetes_data = diabetes_data[diabetes_data.Pregnancies.isin(preg_val_to_delete) == False]


diabetes_data['Insulin'] = diabetes_data['Insulin'].fillna(diabetes_data.Insulin.mean())
print(diabetes_data.Insulin)

[35. 29. nan 23. 32. 45. 19. 47. 38. 30. 41. 33. 26. 15. 36. 11. 31. 37.
 42. 25. 18. 24. 39. 27. 21. 34. 10. 60. 13. 20. 22. 28. 54. 40. 51. 56.
 14. 17. 50. 44. 12. 46. 16.  7. 52. 43. 48.  8. 49. 63.]
0      155.69821
1      155.69821
2      155.69821
3       94.00000
4      168.00000
         ...    
763    180.00000
764    155.69821
765    112.00000
766    155.69821
767    155.69821
Name: Insulin, Length: 763, dtype: float64
